render/naga-agal/src/builder.rs

```diff
@@ -808,43 +808,115 @@ impl<'a> NagaBuilder<'a> {
+                // register (e.g. 'va0'), and 'X' is a component ('X, 'Y', Z', or 'W').
+                // Currently, we only support this when the 'outer' (non-index) register is
+                // a constant register, since we always access constant registers through
+                // an array access.
+                match source.register_type {
+                    RegisterType::Constant => {
+                        // Load the index register (e.g. 'va0') as normal, and access the component
+                        // given by 'index_select'. This is 'va0.x' in the above example.
+                        let (base_index, _format) =
+                            load_register(&source.index_type, source.reg_num)?;
+                        let index_expr = self.evaluate_expr(Expression::AccessIndex {
+                            base: base_index,
+                            index: source.index_select as u32,
+                        });
+
+                        // Convert to an integer, since we're going to be indexing an array
+                        let index_integer = self.evaluate_expr(Expression::As {
+                            expr: index_expr,
+                            kind: ScalarKind::Uint,
+                            convert: Some(4),
+                        });
+
+                        let offset_constant = self.module.constants.append(
+                            Constant {
+                                name: None,
+                                specialization: None,
+                                inner: ConstantInner::Scalar {
+                                    width: 4,
+                                    value: ScalarValue::Uint(source.indirect_offset as u64),
+                                },
+                            },
+                            Span::UNDEFINED,
+                        );
+                        let offset_constant = self
+                            .func
+                            .expressions
+                            .append(Expression::Constant(offset_constant), Span::UNDEFINED);
+
+                        // Add the offset to the loaded value. THis gives us `va0.x + offset` in the above example.
+                        let index_with_offset = self.evaluate_expr(Expression::Binary {
+                            op: BinaryOperator::Add,
+                            left: index_integer,
+                            right: offset_constant,
+                        });
+
+                        let register_pointer = self.func.expressions.append(
+                            Expression::Access {
+                                base: self.constant_registers,
+                                index: index_with_offset,
+                            },
+                            Span::UNDEFINED,
+                        );
+
+                        // Perform the actual load, giving us 'vc[va0.x + offset]' in the above example.
```

Comment:
(Is it correctly understood that `vc` is not actually read, and that the final value `register_pointer` already points to the correct place in "memory"?)

Reply:
We construct a pointer into the `vc` array, and load from it here. We don't load the entire `vc` array (that is, we don't make a copy of it).